src/plugins/data/server/search/session/session_service.ts

```diff
@@ -0,0 +1,205 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { CoreStart, KibanaRequest, SavedObjectsClientContract } from 'kibana/server';
+import {
+  BackgroundSessionSavedObjectAttributes,
+  IKibanaSearchRequest,
+  ISearchOptions,
+  SearchSessionFindOptions,
+  BackgroundSessionStatus,
+} from '../../../common';
+import { BACKGROUND_SESSION_TYPE } from '../../saved_objects';
+import { createRequestHash } from './utils';
+
+const DEFAULT_EXPIRATION = 7 * 24 * 60 * 60 * 1000;
+
+export interface BackgroundSessionDependencies {
+  savedObjectsClient: SavedObjectsClientContract;
+}
+
+export type ISearchSessionClient = ReturnType<
+  ReturnType<BackgroundSessionService['asScopedProvider']>
+>;
+
+export class BackgroundSessionService {
+  /**
+   * Map of sessionId to { [requestHash]: searchId }
+   * @private
+   */
+  private sessionSearchMap = new Map<string, Map<string, string>>();
+
+  constructor() {}
+
+  public setup = () => {};
+
+  public start = (core: CoreStart) => {
+    return {
+      asScoped: this.asScopedProvider(core),
+    };
+  };
+
+  public stop = () => {
+    this.sessionSearchMap.clear();
+  };
+
+  // TODO: Generate the `userId` from the realm type/realm name/username
+  public save = async (
+    sessionId: string,
+    name: string,
+    url: string,
+    expires: Date = new Date(Date.now() + DEFAULT_EXPIRATION),
```

Comment:
There could be multiple requests/responses for a single session. What we should actually probably be doing here is actually calling the `extend` function (when it is implemented) to set the `expires` in each search to this value, rather than the other way around. Thoughts?